### PR TITLE
HADOOP-17267: Add debug-level logs in Filesystem#close

### PR DESCRIPTION
**Pull request by klcopp**
HDFS reuses the same cached FileSystem object across the file system. If the client calls FileSystem.close(), closeAllForUgi(), or closeAll() (if it applies to the instance) anywhere in the system it purges the cache of that FS instance, and trying to use the instance results in an IOException: FileSystem closed.

It would be a great help to clients to see where and when a given FS instance was closed. I.e. in close(), closeAllForUgi(), or closeAll(), it would be great to see a DEBUG-level log of
- calling method name, class, file name/line number
- FileSystem object's identity hash (FileSystem.close() only)

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 22s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  30m 14s |  trunk passed  |
| +1 :green_heart: |  compile  |  23m 36s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |  21m  2s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   1m  0s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 40s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  19m 31s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 37s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 38s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +0 :ok: |  spotbugs  |   2m 38s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   2m 29s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 59s |  the patch passed  |
| +1 :green_heart: |  compile  |  23m 31s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javac  |  23m 31s |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 23s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  javac  |  20m 23s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 50s |  hadoop-common-project/hadoop-common: The patch generated 4 new + 76 unchanged - 0 fixed = 80 total (was 76)  |
| +1 :green_heart: |  mvnsite  |   1m 34s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  17m 28s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 38s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 38s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  findbugs  |   2m 39s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  10m 39s |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  0s |  The patch does not generate ASF License warnings.  |
|  |   | 187m 12s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2321/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2321 |
| JIRA Issue | HADOOP-17267 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 3ed3d0909910 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 7a6265ac425 |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2321/1/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2321/1/testReport/ |
| Max. process+thread count | 1366 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2321/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:625]: pass in "" instead of null and theres' no need for the ? : below

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:639]: just use ugi and let the automatic to.String to the work

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:647]: Prefer SLF4J {} and the values as varargs.
You might want to think that at TRACE the entire throwable is logged. Why so? one level up isn't always enough

**Review by steveloughran — CHANGES_REQUESTED** [review]
patch looks good, some minor changes.

No need for a test, but I would like to see what the actual output looks like

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (klcopp) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:647]: Hi Steve, thanks for taking a look at this!
Could you clarify what you mean by this?: You might want to think that at TRACE the entire throwable is logged. Why so? one level up isn't always enough

**Comment by steveloughran** [comment]
Imagine the closeAllForUGI is wrapped by some method "cleanup()". The log would say "closed in cleanup", but not what cleanup() was invoked from

**Comment by klcopp** [comment]
@steveloughran  I see... do you mean that if TRACE-level logging is enabled, we should log the whole stack trace? and if the level is set to DEBUG, we should just log the calling method?

**Comment by steveloughran** [comment]
> do you mean that if TRACE-level logging is enabled, we should log the whole stack trace? and if the level is set to DEBUG, we should just log the calling method?

exactly.

**Comment by klcopp** [comment]
@steveloughran  Here's what TRACE-level logs (DEBUG-level included, of course) look like:

`closeAll:

[main] DEBUG org.apache.hadoop.fs.FileSystem  - FileSystem.closeAll() called by method: org.apache.hadoop.fs.TestFileSystem.testFsCache(TestFileSystem.java:525); 
[main] TRACE org.apache.hadoop.fs.FileSystem  - FileSystem.closeAll() full stack trace:
java.lang.Throwable
	at org.apache.hadoop.fs.FileSystem.debugLogFileSystemClose(FileSystem.java:649)
	at org.apache.hadoop.fs.FileSystem.closeAll(FileSystem.java:625)
	at org.apache.hadoop.fs.TestFileSystem.testFsCache(TestFileSystem.java:525)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

close:

[main] DEBUG org.apache.hadoop.fs.FileSystem  - FileSystem.close() called by method: org.apache.hadoop.fs.FilterFileSystem.close(FilterFileSystem.java:524); Key: (foo (auth:SIMPLE))@file://; Object Identity Hash: 35645047
[main] TRACE org.apache.hadoop.fs.FileSystem  - FileSystem.close() full stack trace:
java.lang.Throwable
	at org.apache.hadoop.fs.FileSystem.debugLogFileSystemClose(FileSystem.java:649)
	at org.apache.hadoop.fs.FileSystem.close(FileSystem.java:2590)
	at org.apache.hadoop.fs.FilterFileSystem.close(FilterFileSystem.java:524)
	at org.apache.hadoop.fs.FileSystem$Cache.closeAll(FileSystem.java:3580)
	at org.apache.hadoop.fs.FileSystem$Cache.closeAll(FileSystem.java:3551)
	at org.apache.hadoop.fs.FileSystem.closeAll(FileSystem.java:627)
	at org.apache.hadoop.fs.TestFileSystem.testFsCache(TestFileSystem.java:525)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

closeAllForUGI:

[main] DEBUG org.apache.hadoop.fs.FileSystem  - FileSystem.closeAllForUGI() called by method: org.apache.hadoop.fs.TestFileSystemCaching.testCloseAllForUGI(TestFileSystemCaching.java:231); UGI: foo (auth:SIMPLE)
[main] TRACE org.apache.hadoop.fs.FileSystem  - FileSystem.closeAllForUGI() full stack trace:
java.lang.Throwable
	at org.apache.hadoop.fs.FileSystem.debugLogFileSystemClose(FileSystem.java:649)
	at org.apache.hadoop.fs.FileSystem.closeAllForUGI(FileSystem.java:639)
	at org.apache.hadoop.fs.TestFileSystemCaching.testCloseAllForUGI(TestFileSystemCaching.java:231)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
`

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:649]: just use the throwable created on Line 645

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java:646]: * include URI of filesystem
* pass in callingMethod and let it's toString() include all the information; no need to replicate -especially as toString handles file/line unknown.

**Review by steveloughran — CHANGES_REQUESTED** [review]
thanks for the trace, now some bits of tuning

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
Comment (klcopp) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java]: I included the URI in the case of close().

**Comment by steveloughran** [comment]
Perfect. I expect Yetus will be complaining about line width, as there's generally a goal of staying @ 80 chars, but
other than dealing with that, the patch is is ready to go in.

**Comment by steveloughran** [comment]
LGTM, +1 pending yetus

I think we should mention this in the release notes -at least the JIRA text should say the name of the log to set to debug to see this, trace for the full stack. Something which can be done after the PR is merged

**Review by liuml07 — APPROVED** [review]
Is it better to move `if (LOGGER.isDebugEnabled()) {` into  `debugLogFileSystemClose()`?

Otherwise +1

**Comment by klcopp** [comment]
@liuml07  I wanted to add the minimum amount of cost possible.

**Comment by klcopp** [comment]
@steveloughran @liuml07  is there anything else that needs changing?

**Comment by liuml07** [comment]
>  add the minimum amount of cost possible.

How leaving `if (LOGGER.isDebugEnabled())` out of `debugLogFileSystemClose()` do that? Saving one function call? I do not assume that is a concern. I'm more concerned if people forget to check that when calling the method, so make it in the function is preferred to me.

**Comment by steveloughran** [comment]
we could go round in circles worrying about nuances here. Key thing: better insight into when FS instances were closed can help debug some issues, and while there are now a couple of operations executed even when the log is off, they are minimal.

Merging as is

**Comment by klcopp** [comment]
Sweet, thanks!

**Pull request merged**
